tests/form_params_test.go

```diff
@@ -0,0 +1,269 @@
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+}
+
+// Ensure passing a negative Content-Length or one that overflows results in a
+// 400 Bad Request
+func TestFormParametersBadContentLength(t *testing.T) {
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: -1" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+
+	postReq = []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: 10000000000000000000000000" + reqBody + "\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+}
+
+// Tests behaviour when multiple Content-Length are passed. If they are
+// duplicated and of equal
+// length, the request will succeed, otherwise it fails. This conforms to RFC
+// 7320, Section 3.3.2 which says that the server can either replace duplicate
+// Content-Length headers with one value or reject the request. However, for a
+// list of identical values the request is rejected, which is rather inconsistent.
+func TestFormParametersDuplicateContentLength(t *testing.T) {
+	t.Skip()
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n" + "Content-Length: 13\r\n" + "Content-Length: 13\r\n" + "\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
+			t.Errorf("Expected Content-Length to be 13 but got %s", contentLen)
+			if val := req.FormValue("veggie"); val != "potato" {
+				t.Errorf("req.FormValue: want potato, got %s", val)
+			}
```

Comment:
Why test this? We already know the test has failed.

Reply:
We prefer to "keep going" with checks, unless an error makes further checks always fail.

Reply:
But this test is only performed if the test has already failed. It is **not** performed if the previous check succeeded. 
```
if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
	t.Errorf("Expected Content-Length to be 13 but got %s", contentLen)
	if val := req.FormValue("veggie"); val != "potato" {
		t.Errorf("req.FormValue: want potato, got %s", val)	
	}
}
```
See what I mean?

Reply:
Ah, yeah thanks for pointing out, the second if-statement should be outside.